diff --git a/tools/driver/CMakeLists.txt b/tools/driver/CMakeLists.txt
index 805aebf..a5ec9cd 100644
--- a/tools/driver/CMakeLists.txt
+++ b/tools/driver/CMakeLists.txt
@@ -38,7 +38,7 @@ target_link_libraries(clang
   clangFrontendTool
   )
 
-set_target_properties(clang PROPERTIES VERSION ${CLANG_EXECUTABLE_VERSION})
+# set_target_properties(clang PROPERTIES VERSION ${CLANG_EXECUTABLE_VERSION})
 
 # Support plugins.
 if(CLANG_PLUGIN_SUPPORT)
@@ -47,14 +47,14 @@ endif()
 
 add_dependencies(clang clang-headers)
 
-if(UNIX)
+# if(UNIX)
   set(CLANGXX_LINK_OR_COPY create_symlink)
 # Create a relative symlink
   set(clang_binary "clang${CMAKE_EXECUTABLE_SUFFIX}")
-else()
+# else()
   set(CLANGXX_LINK_OR_COPY copy)
   set(clang_binary "${LLVM_RUNTIME_OUTPUT_INTDIR}/clang${CMAKE_EXECUTABLE_SUFFIX}")
-endif()
+# endif()
 
 # Create the clang++ symlink in the build directory.
 set(clang_pp "${LLVM_RUNTIME_OUTPUT_INTDIR}/clang++${CMAKE_EXECUTABLE_SUFFIX}")
